AdvLoggerPkg/Library/AdvancedLoggerAccessLib/AdvancedLoggerAccessLib.c

```diff
@@ -128,22 +180,43 @@ AdvancedLoggerAccessLibGetNextMessageBlock (
 
   if (BlockEntry->Message == NULL) {
     LogEntry = (ADVANCED_LOGGER_MESSAGE_ENTRY *)PTR_FROM_PA (mLoggerInfo->LogBuffer);
+    if (LogEntry->Signature == MESSAGE_ENTRY_SIGNATURE_V2) {
+      // This is actually a v2 entry.
+      LogEntryV2 = (ADVANCED_LOGGER_MESSAGE_ENTRY_V2 *)LogEntry;
+    }
   } else {
     LogEntry = (ADVANCED_LOGGER_MESSAGE_ENTRY *)MESSAGE_ENTRY_FROM_MSG (BlockEntry->Message);
     if (LogEntry->Signature != MESSAGE_ENTRY_SIGNATURE) {
-      DEBUG ((DEBUG_ERROR, "Resume LogEntry invalid signature at %p\n", LogEntry));
-      DUMP_HEX (DEBUG_INFO, 0, (CHAR8 *)LogEntry - 128, 256, "");
-      return EFI_INVALID_PARAMETER;
+      // If this is not a v1 entry, this might be a v2 entry.
+      LogEntryV2 = (ADVANCED_LOGGER_MESSAGE_ENTRY_V2 *)MESSAGE_ENTRY_FROM_MSG_V2 (BlockEntry->Message, BlockEntry->MessageOffset);
+      if (LogEntryV2->Signature != MESSAGE_ENTRY_SIGNATURE_V2) {
+        DEBUG ((DEBUG_ERROR, "Resume LogEntry invalid signature at %p or %p\n", LogEntry, LogEntryV2));
+        DUMP_HEX (DEBUG_INFO, 0, (CHAR8 *)LogEntry - 128, 256, "");
+        return EFI_INVALID_PARAMETER;
+      }
+    }
+
+    if (LogEntryV2) {
+      LogEntryV2 = NEXT_LOG_ENTRY_V2 (LogEntryV2);
+    } else {
+      LogEntry = NEXT_LOG_ENTRY (LogEntry);
     }
+  }
 
-    LogEntry = NEXT_LOG_ENTRY (LogEntry);
+  // At this point, if LogEntryV2 is not NULL, it points to the next entry to be read.
+  // Otherwise LogEntry will contain the next entry. So we simplify the logic by only
+  // always start from LogEntry. However, note that regardless of how we inherit the pointer

```

Comment:
The line: `So we simplify the logic by only always start from LogEntry.` is confusing.  I'd recommend rewording to something like this: `So we simplify the logic by only using LogEntry and overwriting it to use the LogEntryV2 data as necessary.`

Reply:
Updated. Thanks.